File: cpp/scheduler.cpp
/**
 * CPU Scheduling Engine
 * 
 * Implements CPU scheduling algorithms and outputs execution timeline as JSON.
 * Currently supports: Round Robin, FCFS, SJF (Non-Preemptive), Priority (Non-Preemptive)
 * 
 * Usage:
 *   ./scheduler                          → Round Robin with default processes
 *   ./scheduler RR 3 "1,0,5;2,1,3;3,2,8"  → Round Robin, quantum=3, custom processes
 *   ./scheduler FCFS 0 "1,0,5;2,1,3"      → FCFS with custom processes
 *   ./scheduler SJF 0 "1,0,5;2,1,3"       → SJF with custom processes
 *   ./scheduler PRIORITY 0 "1,0,5,2;2,1,3,1" → Priority with custom processes (lower = higher priority)
 * 
 * Process format: "pid,arrivalTime,burstTime[,priority];..."
 * Output: JSON array of { pid, start, end } timeline blocks
 */

#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <sstream>
#include <algorithm>
#include <climits>

struct Process {
    int pid;
    int arrivalTime;
    int burstTime;
    int remainingTime;
    int priority;       // Lower value = higher priority
    int completionTime;
    int turnaroundTime;
    int waitingTime;
};

struct TimeBlock {
    int pid;
    int start;
    int end;
};

// ─── Scheduling Algorithms ────────────────────────────────────────────────────

/**
 * Round Robin Scheduling
 * Preemptive algorithm with fixed time quantum.
 */
std::vector<TimeBlock> roundRobin(std::vector<Process> processes, int quantum) {
    std::vector<TimeBlock> timeline;
    std::queue<int> readyQueue; // indices into processes
    int currentTime = 0;
    int completed = 0;
    int n = processes.size();
    std::vector<bool> inQueue(n, false);

    // Sort by arrival time for initial ordering
    std::sort(processes.begin(), processes.end(),
        [](const Process& a, const Process& b) {
            return a.arrivalTime < b.arrivalTime;
        });

    // Enqueue processes that arrive at time 0
    for (int i = 0; i < n; i++) {
        if (processes[i].arrivalTime <= 0) {
            readyQueue.push(i);
            inQueue[i] = true;
        }
    }

    while (completed < n) {
        if (readyQueue.empty()) {
            // CPU idle — advance to next arriving process
            int nextArrival = INT_MAX;
            for (int i = 0; i < n; i++) {
                if (processes[i].remainingTime > 0 && processes[i].arrivalTime < nextArrival) {
                    nextArrival = processes[i].arrivalTime;
                }
            }
            currentTime = nextArrival;
            for (int i = 0; i < n; i++) {
                if (!inQueue[i] && processes[i].arrivalTime <= currentTime && processes[i].remainingTime > 0) {
                    readyQueue.push(i);
                    inQueue[i] = true;
                }
            }
            continue;
        }

        int idx = readyQueue.front();
        readyQueue.pop();

        int execTime = std::min(quantum, processes[idx].remainingTime);
        int startTime = currentTime;
        currentTime += execTime;
        processes[idx].remainingTime -= execTime;

        // Record timeline block
        timeline.push_back({processes[idx].pid, startTime, currentTime});

        // Enqueue newly arrived processes (arrived during this quantum)
        for (int i = 0; i < n; i++) {
            if (!inQueue[i] && processes[i].arrivalTime <= currentTime && processes[i].remainingTime > 0) {
                readyQueue.push(i);
                inQueue[i] = true;
            }
        }

        // If current process not finished, re-enqueue it
        if (processes[idx].remainingTime > 0) {
            readyQueue.push(idx);
        } else {
            completed++;
            processes[idx].completionTime = currentTime;
            processes[idx].turnaroundTime = currentTime - processes[idx].arrivalTime;
            processes[idx].waitingTime = processes[idx].turnaroundTime - processes[idx].burstTime;
        }
    }

    return timeline;
}

/**
 * First Come First Served (FCFS) Scheduling
 * Non-preemptive, processes execute in arrival order.
 */
std::vector<TimeBlock> fcfs(std::vector<Process> processes) {
    std::vector<TimeBlock> timeline;
    int n = processes.size();

    // Sort by arrival time
    std::sort(processes.begin(), processes.end(),
        [](const Process& a, const Process& b) {
            return a.arrivalTime < b.arrivalTime;
        });

    int currentTime = 0;
    for (int i = 0; i < n; i++) {
        if (currentTime < processes[i].arrivalTime) {
            currentTime = processes[i].arrivalTime;
        }
        int startTime = currentTime;
        currentTime += processes[i].burstTime;
        timeline.push_back({processes[i].pid, startTime, currentTime});
    }

    return timeline;
}

/**
 * Shortest Job First (SJF) - Non-Preemptive
 * Selects the process with shortest burst time among arrived processes.
 */
std::vector<TimeBlock> sjf(std::vector<Process> processes) {
    std::vector<TimeBlock> timeline;
    int n = processes.size();
    int completed = 0;
    int currentTime = 0;
    std::vector<bool> done(n, false);

    while (completed < n) {
        int shortest = -1;
        int minBurst = INT_MAX;

        for (int i = 0; i < n; i++) {
            if (!done[i] && processes[i].arrivalTime <= currentTime && processes[i].burstTime < minBurst) {
                minBurst = processes[i].burstTime;
                shortest = i;
            }
        }

        if (shortest == -1) {
            // No process available — advance time
            int nextArrival = INT_MAX;
            for (int i = 0; i < n; i++) {
                if (!done[i] && processes[i].arrivalTime < nextArrival) {
                    nextArrival = processes[i].arrivalTime;
                }
            }
            currentTime = nextArrival;
            continue;
        }

        int startTime = currentTime;
        currentTime += processes[shortest].burstTime;
        timeline.push_back({processes[shortest].pid, startTime, currentTime});
        done[shortest] = true;
        completed++;
    }

    return timeline;
}

/**
 * Priority Scheduling - Non-Preemptive
 * Lower priority value = higher priority.
 */
std::vector<TimeBlock> priorityScheduling(std::vector<Process> processes) {
    std::vector<TimeBlock> timeline;
    int n = processes.size();
    int completed = 0;
    int currentTime = 0;
    std::vector<bool> done(n, false);

    while (completed < n) {
        int best = -1;
        int bestPriority = INT_MAX;

        for (int i = 0; i < n; i++) {
            if (!done[i] && processes[i].arrivalTime <= currentTime) {
                if (processes[i].priority < bestPriority ||
                    (processes[i].priority == bestPriority && processes[i].arrivalTime < (best >= 0 ? processes[best].arrivalTime : INT_MAX))) {
                    bestPriority = processes[i].priority;
                    best = i;
                }
            }
        }

        if (best == -1) {
            int nextArrival = INT_MAX;
            for (int i = 0; i < n; i++) {
                if (!done[i] && processes[i].arrivalTime < nextArrival) {
                    nextArrival = processes[i].arrivalTime;
                }
            }
            currentTime = nextArrival;
            continue;
        }

        int startTime = currentTime;
        currentTime += processes[best].burstTime;
        timeline.push_back({processes[best].pid, startTime, currentTime});
        done[best] = true;
        completed++;
    }

    return timeline;
}

// ─── Input Parsing ────────────────────────────────────────────────────────────

/**
 * Parse process string in format: "pid,at,bt[,priority];pid,at,bt[,priority];..."
 */
std::vector<Process> parseProcesses(const std::string& input) {
    std::vector<Process> processes;
    std::stringstream ss(input);
    std::string processStr;

    while (std::getline(ss, processStr, ';')) {
        if (processStr.empty()) continue;

        std::stringstream ps(processStr);
        std::string token;
        std::vector<int> values;

        while (std::getline(ps, token, ',')) {
            values.push_back(std::stoi(token));
        }

        if (values.size() >= 3) {
            Process p;
            p.pid = values[0];
            p.arrivalTime = values[1];
            p.burstTime = values[2];
            p.remainingTime = p.burstTime;
            p.priority = (values.size() >= 4) ? values[3] : 0;
            p.completionTime = 0;
            p.turnaroundTime = 0;
            p.waitingTime = 0;
            processes.push_back(p);
        }
    }

    return processes;
}

// ─── JSON Output ──────────────────────────────────────────────────────────────

/**
 * Convert timeline to JSON string.
 * Outputs ONLY valid JSON to stdout — no debug/log messages.
 */
std::string toJSON(const std::vector<TimeBlock>& timeline) {
    std::string json = "[\n";
    for (size_t i = 0; i < timeline.size(); i++) {
        json += "  { \"pid\": " + std::to_string(timeline[i].pid)
             + ", \"start\": " + std::to_string(timeline[i].start)
             + ", \"end\": " + std::to_string(timeline[i].end) + " }";
        if (i < timeline.size() - 1) json += ",";
        json += "\n";
    }
    json += "]";
    return json;
}

// ─── Main Entry Point ─────────────────────────────────────────────────────────

int main(int argc, char* argv[]) {
    std::string algorithm = "RR";
    int quantum = 2;
    std::string processInput = "";

    // Parse command-line arguments
    if (argc >= 2) algorithm = argv[1];
    if (argc >= 3) quantum = std::stoi(argv[2]);
    if (argc >= 4) processInput = argv[3];

    // Default processes if none provided
    std::vector<Process> processes;
    if (processInput.empty()) {
        processes = {
            {1, 0, 5, 5, 3, 0, 0, 0},
            {2, 1, 3, 3, 1, 0, 0, 0},
            {3, 2, 8, 8, 4, 0, 0, 0},
            {4, 3, 6, 6, 2, 0, 0, 0}
        };
    } else {
        processes = parseProcesses(processInput);
    }

    // Run selected algorithm
    std::vector<TimeBlock> timeline;

    if (algorithm == "RR") {
        timeline = roundRobin(processes, quantum);
    } else if (algorithm == "FCFS") {
        timeline = fcfs(processes);
    } else if (algorithm == "SJF") {
        timeline = sjf(processes);
    } else if (algorithm == "PRIORITY") {
        timeline = priorityScheduling(processes);
    } else {
        // Default to Round Robin
        timeline = roundRobin(processes, quantum);
    }

    // Output ONLY JSON to stdout
    std::cout << toJSON(timeline);

    return 0;
}

File: backend/scheduler.py
"""
CPU Scheduling Engine (Python)

Implements CPU scheduling algorithms and returns execution timelines.
Supports: Round Robin, FCFS, SJF (Non-Preemptive), Priority (Non-Preemptive)

Each algorithm returns a list of {"pid": int, "start": int, "end": int} dicts.
"""

from collections import deque
from dataclasses import dataclass, field


@dataclass
class Process:
    pid: int
    arrival_time: int
    burst_time: int
    priority: int = 0
    remaining_time: int = field(init=False)

    def __post_init__(self):
        self.remaining_time = self.burst_time


def round_robin(processes: list[Process], quantum: int) -> list[dict]:
    """
    Round Robin Scheduling — preemptive with fixed time quantum.
    """
    procs = sorted(processes, key=lambda p: p.arrival_time)
    n = len(procs)
    remaining = [p.burst_time for p in procs]
    in_queue = [False] * n
    ready_queue: deque[int] = deque()
    timeline: list[dict] = []
    current_time = 0
    completed = 0

    # Enqueue processes that arrive at time 0
    for i in range(n):
        if procs[i].arrival_time <= 0:
            ready_queue.append(i)
            in_queue[i] = True

    while completed < n:
        if not ready_queue:
            # CPU idle — advance to next arriving process
            next_arrival = min(
                procs[i].arrival_time
                for i in range(n)
                if remaining[i] > 0
            )
            current_time = next_arrival
            for i in range(n):
                if (not in_queue[i]
                        and procs[i].arrival_time <= current_time
                        and remaining[i] > 0):
                    ready_queue.append(i)
                    in_queue[i] = True
            continue

        idx = ready_queue.popleft()
        exec_time = min(quantum, remaining[idx])
        start_time = current_time
        current_time += exec_time
        remaining[idx] -= exec_time

        timeline.append({
            "pid": procs[idx].pid,
            "start": start_time,
            "end": current_time,
        })

        # Enqueue newly arrived processes (arrived during this quantum)
        for i in range(n):
            if (not in_queue[i]
                    and procs[i].arrival_time <= current_time
                    and remaining[i] > 0):
                ready_queue.append(i)
                in_queue[i] = True

        # Re-enqueue current process if not finished
        if remaining[idx] > 0:
            ready_queue.append(idx)
        else:
            completed += 1

    return timeline


def fcfs(processes: list[Process]) -> list[dict]:
    """
    First Come First Served — non-preemptive, arrival order.
    """
    procs = sorted(processes, key=lambda p: p.arrival_time)
    timeline: list[dict] = []
    current_time = 0

    for p in procs:
        if current_time < p.arrival_time:
            current_time = p.arrival_time
        start_time = current_time
        current_time += p.burst_time
        timeline.append({"pid": p.pid, "start": start_time, "end": current_time})

    return timeline


def sjf(processes: list[Process]) -> list[dict]:
    """
    Shortest Job First — non-preemptive.
    Selects the process with shortest burst time among arrived processes.
    """
    procs = list(processes)
    n = len(procs)
    timeline: list[dict] = []
    done = [False] * n
    current_time = 0
    completed = 0

    while completed < n:
        # Find shortest burst among arrived, undone processes
        shortest = -1
        min_burst = float("inf")

        for i in range(n):
            if (not done[i]
                    and procs[i].arrival_time <= current_time
                    and procs[i].burst_time < min_burst):
                min_burst = procs[i].burst_time
                shortest = i

        if shortest == -1:
            # No process available — advance time
            next_arrival = min(
                procs[i].arrival_time for i in range(n) if not done[i]
            )
            current_time = next_arrival
            continue

        start_time = current_time
        current_time += procs[shortest].burst_time
        timeline.append({
            "pid": procs[shortest].pid,
            "start": start_time,
            "end": current_time,
        })
        done[shortest] = True
        completed += 1

    return timeline


def priority_scheduling(processes: list[Process]) -> list[dict]:
    """
    Priority Scheduling — non-preemptive.
    Lower priority value = higher priority.
    Ties broken by earlier arrival time.
    """
    procs = list(processes)
    n = len(procs)
    timeline: list[dict] = []
    done = [False] * n
    current_time = 0
    completed = 0

    while completed < n:
        best = -1
        best_priority = float("inf")

        for i in range(n):
            if not done[i] and procs[i].arrival_time <= current_time:
                if (procs[i].priority < best_priority
                        or (procs[i].priority == best_priority
                            and procs[i].arrival_time < (
                                procs[best].arrival_time if best >= 0 else float("inf")
                            ))):
                    best_priority = procs[i].priority
                    best = i

        if best == -1:
            next_arrival = min(
                procs[i].arrival_time for i in range(n) if not done[i]
            )
            current_time = next_arrival
            continue

        start_time = current_time
        current_time += procs[best].burst_time
        timeline.append({
            "pid": procs[best].pid,
            "start": start_time,
            "end": current_time,
        })
        done[best] = True
        completed += 1

    return timeline


# ─── Dispatcher ────────────────────────────────────────────────────────────────

ALGORITHMS = {
    "RR": lambda procs, q: round_robin(procs, q),
    "FCFS": lambda procs, _: fcfs(procs),
    "SJF": lambda procs, _: sjf(procs),
    "PRIORITY": lambda procs, _: priority_scheduling(procs),
}


def run_algorithm(algorithm: str, quantum: int, processes: list[Process]) -> list[dict]:
    """Run a scheduling algorithm by name and return the timeline."""
    algo_fn = ALGORITHMS.get(algorithm.upper())
    if algo_fn is None:
        algo_fn = ALGORITHMS["RR"]
    return algo_fn(processes, quantum)
